kie-maven-plugin/src/main/java/org/kie/maven/plugin/GeneratePMMLModelMojo.java

```diff
@@ -0,0 +1,322 @@
+
+    @Parameter(required = true, defaultValue = "${project.basedir}")
+    private File projectDir;
+
+    @Parameter(required = true, defaultValue = "${project.build.testSourceDirectory}")
+    private File testDir;
+
+    @Parameter
+    private Map<String, String> properties;
+
+    @Parameter(required = true, defaultValue = "${project}")
+    private MavenProject project;
+
+    @Parameter(required = true, defaultValue = "${project.build.outputDirectory}")
+    private File outputDirectory;
+
+    @Parameter(property = "generatePMMLModel", defaultValue = "no")
+    private String generatePMMLModel;
+
+    private List<Path> createdPaths = new ArrayList<>();
+    private List<Path> movedPmmlFiles = new ArrayList<>();
+
+    @Override
+    public void execute() throws MojoExecutionException, MojoFailureException {
+        // GeneratePMMLModelMojo is executed when BuildMojo and GenerateModelMojo aren't and vice-versa
+        boolean modelParameterEnabled = PMMLModelMode.modelParameterEnabled(generatePMMLModel);
+        boolean modelCompilerInClassPath = isModelCompilerInClassPath(project.getDependencies());
+        if (modelParameterEnabled && modelCompilerInClassPath) {
+            generateModel();
+        } else if (modelParameterEnabled) { // !modelCompilerInClassPath
+            getLog().warn("You're trying to build rule assets in a project from an executable rule model, but you did not provide the required dependency on the project classpath.\n" +
+                                  "To enable executable rule models for your project, add the `drools-model-compiler` dependency in the `pom.xml` file of your project.\n");
+        }
+    }
+
+    private void generateModel() throws MojoExecutionException {
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            Set<URL> urls = new HashSet<>();
+            for (String element : project.getCompileClasspathElements()) {
+                urls.add(new File(element).toURI().toURL());
+            }
+
+            project.setArtifactFilter(new CumulativeScopeArtifactFilter(Arrays.asList("compile",
+                                                                                      "runtime")));
+            for (Artifact artifact : project.getArtifacts()) {
+                File file = artifact.getFile();
+                if (file != null) {
+                    urls.add(file.toURI().toURL());
+                }
+            }
+            urls.add(outputDirectory.toURI().toURL());
+
+            ClassLoader projectClassLoader = URLClassLoader.newInstance(urls.toArray(new URL[0]),
+                                                                        getClass().getClassLoader());
+            Thread.currentThread().setContextClassLoader(projectClassLoader);
+        } catch (DependencyResolutionRequiredException | MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
What about add a `finally` block to restore contextClassLoader?

Reply:
@danielezonca 
MojoExecutionException completely interrupt plugin execution, killing any thread